File: Grokking the Coding Interview/Pattern - Sliding Window/longestDistinctSubstring.js
const longestDistinctSubstring = (str) => {
  const indexMap = {};
  let maxLength = -Infinity;
  let windowStart = 0;
  for (let windowsEnd = 0; windowsEnd < str.length; windowsEnd += 1) {
    if (str[windowsEnd] in indexMap) {
      windowStart = indexMap[str[windowsEnd]] + 1;
      delete indexMap[str[windowsEnd]];
    }
    indexMap[str[windowsEnd]] = windowsEnd;
    maxLength = Math.max(maxLength, windowsEnd - windowStart + 1)
  }
  return (maxLength === -Infinity) ? 0 : maxLength;
};

console.log(longestDistinctSubstring('aabccbb')); // -> 3
console.log(longestDistinctSubstring('abbbb')); // -> 2
console.log(longestDistinctSubstring('abccde')); // -> 3

File: Grokking the Coding Interview/Pattern - Sliding Window/longest_distinct_substring.py
def longest_substring_distinct_characters(s):
    ind = {}
    max_len = 0
    l = 0
    for r, char in enumerate(s):
        if char in ind:
            l = max(ind[char] + 1, l)
        ind[char] = r
        max_len = max(max_len, r - l + 1)
    return max_len

print(longest_substring_distinct_characters('aabccbb'))
print(longest_substring_distinct_characters('abbbb'))
print(longest_substring_distinct_characters('abccde'))
